[#159133590] Code refactoring

andela-Taiwo opened the pull request:
#### What does this PR do?
- Code refactoring
#### Description of task completed
- add resource controllers for user and admin
- update tests request parameters to use factories
- fix bug the flash messages
#### How should it be tested?
- `rspec`
#### Related PivotalTracker story
 - #159133590

andela-Taiwo committed:
ch(code refactor): code refactoring

- add resource controllers for user and admin
- update tests request parameters to use factories
- fix bug the flash messages
[Finishes #159133590]

andela-Taiwo committed:
(code refactor)

add session controller tests

morriskimani commented on app/app.rb:
```diff
@@ -110,14 +112,13 @@ def html(view)
     check_admin_authentication
     "
-		<p>Added Job: #{ResourceNotification.set(:queue => :default).perform_async}</p>
+		<p>Added Job: #{ResourceNotification.set(queue: :default).perform_async}</p>
```
Let's have this in a view

morriskimani commented on app/controllers/admin_topic_controller.rb:
```diff
@@ -0,0 +1,65 @@
+      def self.registered(app)
+
+        app.before do
+          check_admin_authentication if request.path_info == "/admin/"
```
:+:

morriskimani commented on app/controllers/admin_topic_controller.rb:
```diff
@@ -0,0 +1,65 @@
+          @topic = Topic.new(params[:topic])
+          if @topic.save
+            flash[:success] = "Successfully add a new topic"
+            redirect "/admin/topic"
```
Let's have this outside the if block

morriskimani commented on app/controllers/admin_topic_controller.rb:
```diff
@@ -0,0 +1,65 @@
+          erb :'topics/edit_topic_form' if @topic
+        end
+
+        app.patch "/admin/topic/:id/edit" do
```
Do we need `edit` in the route name? The HTTP method used already gives an indication of the action that takes place

morriskimani commented on app/controllers/admin_topic_controller.rb:
```diff
@@ -0,0 +1,65 @@
+        app.patch "/admin/topic/:id/edit" do
+          @topic = Topic.find_by_id(params[:id])
+          if @topic
+            erb :'topics/edit_topic_form'
```
Why do we need to render this form here?

morriskimani commented on app/controllers/admin_topic_controller.rb:
```diff
@@ -0,0 +1,65 @@
+        end
+
+        app.delete "/admin/topic/:id/delete" do
+          @topic = Topic.find_by_id(params[:id])
```
Let's DRY out this controller by using a before filter to set this instance variable

emugaya commented on app/controllers/resource_controller.rb:
```diff
@@ -1,83 +1,71 @@
-            @resource.description = params[:description] if params.has_key?('description')
-            @resource.url = params[:url] if params.has_key?('url')
-
+            @resource.title = params[:title].strip if params.has_key?("title")
```
Consider using `resource.update_attributes(params)`

morriskimani commented on spec/controllers/session_spec.rb:
```diff
@@ -0,0 +1,46 @@
+  let(:user) { create :user }
+  let(:admin)  { build :admin }
+  after { Warden.test_reset! }
+  def app
```
I think it might be better to have this as a  helper method

morriskimani reviewed (commented):
Good work!
Just some few comments

andela-Taiwo committed:
(refactor)

- add helper for sending mail
- add test for resource notification helper

andela-Taiwo committed:
(code refactor)

fix before filter bug
add airbnb rubocop config
fix rubocop violations

andela-Taiwo committed:
(code refactor)

-fix user session bug on login and signup

andela-Taiwo committed:
(code refactor)

add redirect to signup controller
add rediret to login controller
add tests for the redirect paths

andela-Taiwo committed:
(code refactor)

add flash message for invalid user login details
add test for the flash message

andela-Taiwo committed:
(code refactor)

fix flash error messages

andela-Taiwo committed:
(code refactor)

fix rubocop violations
